### PR TITLE
Reduced randomness

### PR DESCRIPTION
**Pull request by parkercoates**
A small set of changes, including two new flags to control how pipes are displayed. Could be broken up into smaller pull requests if desired.

### CODE REVIEW

**Comment by Foggalong** [comment]
This pull request is confusing me a bit and yeah should probably be split up into smaller pulls. It also seems to introduce a few bugs which is less than ideal.

* I really like the idea of being able to specify a colour with the `-c` flag and I can't believe it's not something that's been thought of before. It's implementation seems cause some problems though:
  * Breaks the `-C` flag which used to give all white pipes and now just uses the last colour used in `-c`
  * Does weird things if you use numbers ≥8
  * Does [weirder things](http://i.imgur.com/dvehG08.png) if you put something that isn't a valid number
  * If multiple `-c` are used but one is invalid, all pipes come out green
  * It isn't documented in the README
* I'm not entirely sure what the `-E` tag does.
  * I get your commit description but it doesn't really apply to `pipes.sh` since it only draws one pipe at a time. Because of that for me running with `-E` just does the same as `-c 2`.
  * It isn't documented in the README or under `--help`
* If I'm understanding 2ffa69a34dfd8501dad78ae41bd4773f85593dbb correctly then the idea is that if I use `-t 2 -t 3` then the pipes will alternate between 2 & 3 every time a new one is drawn (and the same theory for `-c 2 -c 3`). If that's the case then this isn't working, sometimes a style 2 will be followed by another style 2 and not by a style 3.

Anyways, thanks for the pull request! I definitely like some of the ideas in here :)

**Comment by parkercoates** [comment]
Thanks for the feedback. I didn't modify the README at all, because I didn't know if there was interest in the changes, but I should have mentioned that.

* The `-C` option seem broken even without this change set, for me. With it set, `pipes.sh` doesn't output any colour escape codes, but because it doesn't actually reset anything either. So the pipes are drawn with the last used colour, until the `-r` limit is hit at which point they are drawn colourless. Adding a call to `tput reset` into the script startup code seems to fix this. I'll add this.

* I don't see how `-c` could be causing issues with values above 7, since the argument handling code does a range check and ignores any values outside that range. But good catch on passing in random text to it. The average string evaluates to 0 in a numeric context, which the range check was letting through. I will fix that.

* `-E` and the changes to `-t` only really make sense with `-p` set to something greater than 1. You are correct, with `-p1` it doesn't really do anything. Some examples would probably help:
  * `pipes.sh -E -p5`: 5 pipes, each a different, permanent colour
  * `pipes.sh -E -p2 -c2 -c3`: one permanently green pipe and one permanently yellow pipe 
  * `pipes.sh -E -p2 -c2 -c3 -t2`: one permanently green, curvy pipe and one permanently yellow, curvy pipe
  * `pipes.sh -E -p2 -c2 -t2 -c3 -t6`: one permanently green, curvy pipe and one permanently yellow, dotted pipe
  * `pipes.sh -E -p4 -c2 -t2 -c3 -t6`: two permanently green, curvy pipes and two permanently yellow, dotted pipes
  * `pipes.sh -E -p6 -c2 -c3 -t1 -t6 -t8`: 1 green/curvy, 1 yellow/dotted, 1 green/railroad, 1 yellow/curvy, 1 green/dotted, 1 yellow/railroad

  Maybe this is a sign that the `-h` text for this feature could be clearer. Suggestions are welcome.

* No, the intent of 2ffa69a isn't for a single pipe to cycle through the types, but instead for the initial assignment of pipe types to be deterministic, as is used in the examples about.

**Comment by parkercoates** [comment]
- A few small fixes made.
- README documentation added.

**Comment by livibetter** [comment]
LGTM (didn't test) for almost all except the cfa757e (the second commit), I think it should be an option like `-E` in 1cf6c7d, I consider the random starting color is a feature and I do not want deterministic.

----

I think pipes.sh has gone a long way, although I am not entirely sure if this:

```
pipes.sh -E -p6 -c2 -c3 -t1 -t6 -t8
```

is what I want to see, but it's become more complex for sure.  (and overly complicated for me)

File: pipes.sh [hunk 1]
```diff
@@ -117,7 +122,7 @@ while REPLY=; read -t 0.0$((1000/f)) -n 1 2>/dev/null; [[ -z $REPLY ]] ; do
         ((${l[i]}%2)) && ((x[i]+=-${l[i]}+2,1)) || ((y[i]+=${l[i]}-1))
 
         # Loop on edges (change color on loop):
-        ((!NOEDGECHANGE&&(${x[i]}>=w||${x[i]}<0||${y[i]}>=h||${y[i]}<0))) && ((c[i]=RANDOM%8, v[i]=V[${#V[@]}*RANDOM/M]))
+        ((!NOEDGECHANGE&&(${x[i]}>=w||${x[i]}<0||${y[i]}>=h||${y[i]}<0))) && ((c[i]=${C[$((RANDOM%${#C[@]}))]}, v[i]=V[${#V[@]}*RANDOM/M]))
```
Comment (livibetter) [inline pipes.sh]: The outcome isn't uniform, see [my post](http://yjlv.blogspot.com/2013/07/use-bash-random-right-way.html).

It should be like  `# * RANDOM / M`, but if you don't care, well, it's just a color.

**Comment by parkercoates** [comment]
> I consider the random starting color is a feature and I do not want deterministic.

The starting colour isn't actually random at the moment. It's always red. I agree that having the starting colour be random would be nicer.

I'm now rethinking this patch set, and I think I can do all this in a simpler more understandable way. Hopefully I can have an update up soon.

**Comment by livibetter** [comment]
@parkercoates My comment was from the future!  Just kidding, my memory was messed up with [pipesX.sh](https://github.com/pipeseroni/pipesX.sh/blob/988cbca4ed7aca027d0c784f5d09683a929063f5/pipesX.sh#L139).

I love simplicity, by the way, I think setting `CN=${#C[@]}` would increase readability, shorten the lines, might even improve performance.  (leave `${#V[@]}` for another commit, I found the script has a few mismatch coding styles, my bad, actually)

**Comment by parkercoates** [comment]
Okay, so the changes aren't as substantial as I thought they'd be. The order of the commits makes more sense now, though. I tried to order them from least to most controversial. ;) 

If there's a feeling that the `-K` option introduced in 40033f4 is just too complicated or confusing, I completely understand.

File: pipes.sh [hunk 2]
```diff
@@ -114,12 +114,12 @@ while REPLY=; read -t 0.0$((1000/f)) -n 1 2>/dev/null; [[ -z $REPLY ]] ; do
         ((${l[i]}%2)) && ((x[i]+=-${l[i]}+2,1)) || ((y[i]+=${l[i]}-1))
 
         # Loop on edges (change color on loop):
-        ((${x[i]}>=w||${x[i]}<0||${y[i]}>=h||${y[i]}<0)) && ((c[i]=RANDOM%8, v[i]=V[${#V[@]}*RANDOM/M]))
+        ((${x[i]}>=w||${x[i]}<0||${y[i]}>=h||${y[i]}<0)) && ((c[i]=8*RANDOM/M, v[i]=V[${#V[@]}*RANDOM/M]))
```
Comment (livibetter) [inline pipes.sh]: This one is unnecessary, for divisor n == 2^i, that is 32768 % n == 0, the outcome is evenly distributed.

(I was hoping this commit would come separate (from this PR), after coding style fix, and or within refactoring, but it's fine)

Comment (livibetter) [reply pipes.sh]: I am making this as a change request.

**Comment by livibetter** [comment]
@parkercoates I will merge once you make the change.

Off-topic: If you have time, I think you should apply for pipes.sh new maintainer position at pipeseroni/pipeseroni.github.io#3, currently, we have no one to actively maintain pipes.sh.

**Comment by livibetter** [comment]
@parkercoates I've merged this, if you have more to contribute in the future, please keep PR/commit atomic.

**Comment by parkercoates** [comment]
@livibetter, Thanks for the merge! I'm normally a stickler for atomic review requests, so I doubt this will be an issue in the future. 

And speaking of the future, thank you for the invitation to become a maintainer, but I think I lack both the qualifications and the motivation. I know very little about all the terminal control stuff and really only wanted to make a handful of changes so that I could watch two differently coloured pipes battling it out to dominate the screen.

**Comment by livibetter** [comment]
If you are really lack of them, I'd not have asked.  Anyway, they are cool pipes!

**Comment by livibetter** [comment]
@parkercoates I have stolen your idea in that screenshot to make a [video](https://youtu.be/7djAShRCpYE).

**Pull request merged**